cli/src/commands/extensions/state.rs

```diff
@@ -0,0 +1,102 @@
+//! Shared extension state.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::ops::Deref;
+
+use anyhow::{anyhow, Result};
+use deno_runtime::deno_core::OpState;
+use futures::future::BoxFuture;
+use tokio::sync::Mutex;
+
+use crate::commands::extensions::permissions::Permissions;
+use crate::commands::extensions::PhylumApi;
+
+/// Holds either an unawaited, boxed `Future`, or the result of awaiting the
+/// future.
+enum OnceFuture<T: Unpin> {
+    Future(BoxFuture<'static, T>),
+    Awaited(T),
+}
+
+impl<T: Unpin> OnceFuture<T> {
+    fn new(inner: BoxFuture<'static, T>) -> Self {
+        OnceFuture::Future(inner)
+    }
+
+    async fn get(&mut self) -> &T {
+        match *self {
+            OnceFuture::Future(ref mut inner) => {
+                *self = OnceFuture::Awaited(inner.await);
+                match *self {
+                    OnceFuture::Future(..) => unreachable!(),
+                    OnceFuture::Awaited(ref mut inner) => inner,
+                }
+            },
+            OnceFuture::Awaited(ref mut inner) => inner,
+        }
+    }
+}
+
+// XXX: Holding a `Ref`, `RefMut`, or guard across await points, can cause
+// issues when that field is accessed from another extension API
+// method.
+//
+// If a field in [`ExtensionStateInner`] requires mutable access, it should be
+// synchronized with an async-safe `Mutex` or `RwLock`, so an attempted access
+// to the inner value will cause blocking until the initial guard is resolved.
+//
+// This also means that if this guard is held across an await point, all other
+// access will stall until that future is completed. Avoid holding guards across
+// await points, especially when depending on other methods accessing the same
+// lock.
+//
+// Using `RefCell` to avoid locking should only be done when the value is
+// private and never held across await points internally. Prefer [`Cell`] if the
+// inner value only needs to be replaced.
+//
+/// Extension state the APIs have access to.
+pub struct ExtensionStateInner {
+    pub permissions: Permissions,
+
+    api: Mutex<OnceFuture<Result<Rc<PhylumApi>>>>,
+}
+
+impl ExtensionStateInner {
+    pub async fn api(&self) -> Result<Rc<PhylumApi>> {
+        // This mutex guard is only useful for synchronizing mutable access while
+        // awaiting the PhylumApi future. Subsequent access to the API is
+        // immediate and will not hold the Mutex for extended periods of time.
+        let mut guard = self.api.lock().await;
+        Ok(guard.get().await.as_ref().map_err(|e| anyhow!("{:?}", e))?.clone())
```

Comment:
It would probably involve changing the return type here to be `Result<MutexGuard<&PhylumApi>>` or something crazy.... `Rc` is definitely simpler.

Reply:
You don't want to hold a mutex guard though, because having a mutex guard while interacting with the API will mean that no other function can interact with the API for that entire period.

So that's a really bad idea.

Reply:
Honestly just `Cell` might work?

Reply:
Nvm it definitely doesn't. There's a reason we have `Mutex` here and I'm pretty sure it would be difficult to avoid. We need some barrier to synchronize on and the `Mutex` lock provides exactly that.